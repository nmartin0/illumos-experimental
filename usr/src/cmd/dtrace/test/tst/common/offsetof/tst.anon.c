/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2025 Oxide Computer Company
 */

/*
 * This contains a few anonymous structs and unions (nested), that test that the
 * D language offsetof works correctly in these cases.
 */

struct foo {
	int a;
	union {
		int b;
		int c;
		struct {
			int d;
			int e;
			int f;
		};
		int g[3];
	};
	struct {
		int h;
		union {
			int i;
			struct {
				int j;
				union {
					int k;
					struct {
						int l;
						int m;
						union {
							int n;
							struct {
								int o;
							};
						};
						int p;
					};
					int q;
				};
				int r;
			};
			int s;
		};
		int t;
	};
};

struct foo foo;

int
main(void)
{
	return (0);
}
