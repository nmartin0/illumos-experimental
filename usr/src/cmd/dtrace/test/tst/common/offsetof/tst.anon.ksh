#!/usr/bin/ksh
#
# This file and its contents are supplied under the terms of the
# Common Development and Distribution License ("CDDL"), version 1.0.
# You may only use this file in accordance with the terms of version
# 1.0 of the CDDL.
#
# A full copy of the text of the CDDL should have accompanied this
# source.  A copy of the CDDL is also available via the Internet at
# http://www.illumos.org/license/CDDL.
#

#
# Copyright 2025 Oxide Computer Company
#

#
# Test that we can correctly determine the offsets in various anonymous
# structures and unions.
#

if (( $# != 1 )); then
        printf "%s\n" "expected one argument: <dtrace-path>" >&2
        exit 2
fi

dtrace=$1
$dtrace -c ./tst.anon.exe -qs /dev/stdin <<EOF
BEGIN
{
	printf("a: %d, exp 0\n", offsetof(struct pid\`foo, a));
	printf("b: %d, exp 4\n", offsetof(struct pid\`foo, b));
	printf("c: %d, exp 4\n", offsetof(struct pid\`foo, c));
	printf("d: %d, exp 4\n", offsetof(struct pid\`foo, d));
	printf("e: %d, exp 8\n", offsetof(struct pid\`foo, e));
	printf("f: %d, exp 12\n", offsetof(struct pid\`foo, f));
	printf("g: %d, exp 4\n", offsetof(struct pid\`foo, g));
	printf("h: %d, exp 16\n", offsetof(struct pid\`foo, h));
	printf("i: %d, exp 20\n", offsetof(struct pid\`foo, i));
	printf("j: %d, exp 20\n", offsetof(struct pid\`foo, j));
	printf("k: %d, exp 24\n", offsetof(struct pid\`foo, k));
	printf("l: %d, exp 24\n", offsetof(struct pid\`foo, l));
	printf("m: %d, exp 28\n", offsetof(struct pid\`foo, m));
	printf("n: %d, exp 32\n", offsetof(struct pid\`foo, n));
	printf("o: %d, exp 32\n", offsetof(struct pid\`foo, o));
	printf("p: %d, exp 36\n", offsetof(struct pid\`foo, p));
	printf("q: %d, exp 24\n", offsetof(struct pid\`foo, q));
	printf("r: %d, exp 40\n", offsetof(struct pid\`foo, r));
	printf("s: %d, exp 20\n", offsetof(struct pid\`foo, s));
	printf("t: %d, exp 44\n", offsetof(struct pid\`foo, t));

	exit(0);
}
EOF
