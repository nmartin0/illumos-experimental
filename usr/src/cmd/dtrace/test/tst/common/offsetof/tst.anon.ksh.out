a: 0, exp 0
b: 4, exp 4
c: 4, exp 4
d: 4, exp 4
e: 8, exp 8
f: 12, exp 12
g: 4, exp 4
h: 16, exp 16
i: 20, exp 20
j: 20, exp 20
k: 24, exp 24
l: 24, exp 24
m: 28, exp 28
n: 32, exp 32
o: 32, exp 32
p: 36, exp 36
q: 24, exp 24
r: 40, exp 40
s: 20, exp 20
t: 44, exp 44

